File: CS/CalculatedPropertiesSolution.Mobile/CorsSupport.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace CalculatedPropertiesSolution.Mobile {

    // NOTE
    // The following change to web.config is required
    // <system.serviceModel>
    //    <serviceHostingEnvironment aspNetCompatibilityEnabled="true" /> 

    static class CorsSupport {

        public static void HandlePreflightRequest(HttpContext context) {
            var req = context.Request;
            var res = context.Response;
            var origin = req.Headers["Origin"];

            if (!String.IsNullOrEmpty(origin)) {
                res.AddHeader("Access-Control-Allow-Origin", origin);
                res.AddHeader("Access-Control-Allow-Credentials", "true");
                res.AddHeader("Vary", "Origin");

                var methods = req.Headers["Access-Control-Request-Method"];
                var headers = req.Headers["Access-Control-Request-Headers"];

                if (!String.IsNullOrEmpty(methods))
                    res.AddHeader("Access-Control-Allow-Methods", methods);

                if (!String.IsNullOrEmpty(headers))
                    res.AddHeader("Access-Control-Allow-Headers", headers);

                if (req.HttpMethod == "OPTIONS") {
                    res.StatusCode = 204;
                    res.End();
                }
            }
        }

    }
}

File: VB/CalculatedPropertiesSolution.Mobile/CorsSupport.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web

Namespace CalculatedPropertiesSolution.Mobile

    ' NOTE
    ' The following change to web.config is required
    ' <system.serviceModel>
    '    <serviceHostingEnvironment aspNetCompatibilityEnabled="true" /> 

    Friend NotInheritable Class CorsSupport

        Private Sub New()
        End Sub


        Public Shared Sub HandlePreflightRequest(ByVal context As HttpContext)
            Dim req = context.Request
            Dim res = context.Response
            Dim origin = req.Headers("Origin")

            If Not String.IsNullOrEmpty(origin) Then
                res.AddHeader("Access-Control-Allow-Origin", origin)
                res.AddHeader("Access-Control-Allow-Credentials", "true")
                res.AddHeader("Vary", "Origin")

                Dim methods = req.Headers("Access-Control-Request-Method")
                Dim headers = req.Headers("Access-Control-Request-Headers")

                If Not String.IsNullOrEmpty(methods) Then
                    res.AddHeader("Access-Control-Allow-Methods", methods)
                End If

                If Not String.IsNullOrEmpty(headers) Then
                    res.AddHeader("Access-Control-Allow-Headers", headers)
                End If

                If req.HttpMethod = "OPTIONS" Then
                    res.StatusCode = 204
                    res.End()
                End If
            End If
        End Sub

    End Class
End Namespace
